launch/launch/substitutions/boolean_substitution.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2022 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for boolean substitutions."""
+
+from typing import Iterable
+from typing import Text
+
+from .launch_configuration import LaunchConfiguration
+from .substitution_failure import SubstitutionFailure
+from ..frontend import expose_substitution
+from ..launch_context import LaunchContext
+from ..some_substitutions_type import SomeSubstitutionsType
+from ..substitution import Substitution
+from ..utilities import normalize_to_list_of_substitutions
+from ..utilities.type_utils import perform_typed_substitution
+
+
+@expose_substitution('not')
+class NotSubstitution(Substitution):
+    """Substitution that returns 'not' of the input boolean value."""
+
+    def __init__(self, value: SomeSubstitutionsType) -> None:
+        """Create a NotSubstitution substitution."""
+        super().__init__()
+
+        self.__value = normalize_to_list_of_substitutions(value)
+
+    @classmethod
+    def parse(cls, data: Iterable[SomeSubstitutionsType]):
+        """Parse `NotSubstitution` substitution."""
+        if len(data) != 1:
+            raise TypeError('not substitution expects 1 argument')
+        return cls, {'value': data[0]}
+
+    @property
+    def value(self) -> Substitution:
+        """Getter for value."""
+        return self.__value
+
+    def describe(self) -> Text:
+        """Return a description of this substitution as a string."""
+        return f'NotSubstitution({self.value})'
+
+    def perform(self, context: LaunchContext) -> Text:
+        """Perform the substitution."""
+        try:
+            value = LaunchConfiguration(self.value, default=self.value).perform(context)
```

Comment:
@kenji-miyake hmm, so you're potentially expanding another `LaunchConfiguration` and hoping the result is not itself a valid launch configuration name for the default to kick in? Same as in https://github.com/ros2/launch/pull/600#pullrequestreview-907258188, IMO if you want these substitutions to have access to launch configurations as variables, you have to expand all substitutions first and then attempt a variable lookup. 

CC @ivanpauno @wjwwood

Reply:
> so you're potentially expanding another LaunchConfiguration and hoping the result is not itself a valid launch configuration name for the default to kick in?

@hidmic It was a hacky way. I'm sorry for confusing you. :bow: 

What I wanted to do are:
- Convert `value` =(`TextSubstitution('value')`) to `LaunchConfiguration('value')`.
- Fall back `$(var value)` (= `LaunchConfiguration('value')`).

> Same as in https://github.com/ros2/launch/pull/600#pullrequestreview-907258188, IMO if you want these substitutions to have access to launch configurations as variables, you have to expand all substitutions first and then attempt a variable lookup.

Since it was an optional feature for me, I've dropped it for now.
If you kindly show me some examples, I can work on the feature in a proper way.